### PR TITLE
cl: add death constraints to simple transfer scenario

### PR DESCRIPTION
**Pull request by davidrusu**
We now prove and verify no-op death constraints on the transfer scenario

### CODE REVIEW

File: goas/cl/ledger/src/death_constraint.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    // resisitant mapping of RISC0 ids to cl death constraints.
+
+    let mut hasher = Sha256::new();
+    hasher.update(b"NOMOS_RISC0_ID_TO_CL_DEATH_CONSTRAINT");
+    for word in risc0_id {
+        hasher.update(u32::to_ne_bytes(word));
+    }
+    let death_constraint: [u8; 32] = hasher.finalize().into();
+    death_constraint
+}
```
Comment (zeegomo) [inline goas/cl/ledger/src/death_constraint.rs:26]: can't we just re-interpret the bytes? It's 32 bytes anyway

File: goas/cl/ledger/src/partial_tx.rs [hunk 2]
```diff
@@ -1,26 +1,65 @@
...
+pub struct PartialTxInput {
+    pub input: ProvedInput,
+    pub death: DeathProof,
+}
+
+impl PartialTxInput {
+    fn verify(&self, ptx_root: cl::PtxRoot) -> bool {
+        let nf = self.input.input.input.nullifier;
+        self.input.input.input.death_cm == self.death.death_commitment() // ensure the death proof is actually for this input
+            && self.input.verify() // ensure the input proof is valie
```
Comment (zeegomo) [inline goas/cl/ledger/src/partial_tx.rs:17]: typo: valid

File: goas/cl/ledger/src/partial_tx.rs [hunk 3]
```diff
@@ -1,26 +1,65 @@
...
+#[derive(Debug, Clone)]
+pub struct PartialTxInput {
+    pub input: ProvedInput,
+    pub death: DeathProof,
+}
+
+impl PartialTxInput {
+    fn verify(&self, ptx_root: cl::PtxRoot) -> bool {
+        let nf = self.input.input.input.nullifier;
+        self.input.input.input.death_cm == self.death.death_commitment() // ensure the death proof is actually for this input
```
Comment (zeegomo) [inline goas/cl/ledger/src/partial_tx.rs:16]: where do we check that the `death_cm` is consistent with the note commitment? in the input proof?

File: goas/cl/ledger/src/death_constraint.rs (hunk 1 above)
Comment (davidrusu) [reply goas/cl/ledger/src/death_constraint.rs:26]: yes, that's how I had it initially, but I think it's just a pun that the risc0 stark proof ids happen to be 32 bytes. When we switch to groth16, the constraint vk will not be 32 bytes, and instead we'll need to hash the vk in the same way to bind the note to this vk.

File: goas/cl/ledger/src/partial_tx.rs (hunk 3 above)
Comment (davidrusu) [reply goas/cl/ledger/src/partial_tx.rs:16]: yes, the input proof, through the `InputWitness::commit()` call: https://github.com/logos-co/nomos-pocs/blob/master/goas/cl/cl/src/input.rs#L61

File: goas/cl/ledger/src/death_constraint.rs (hunk 1 above)
Comment (zeegomo) [reply goas/cl/ledger/src/death_constraint.rs:26]: Agree, ideally, all proofs should be groth16 even in the PoC?

**Review by zeegomo — APPROVED** [review]

File: goas/cl/ledger/src/death_constraint.rs (hunk 1 above)
Comment (davidrusu) [reply goas/cl/ledger/src/death_constraint.rs:26]: yeah, ideally. Hopefully its a straight forward change to swap out the stark proofs for groth16 proofs.


Maybe we still want to maintain a full stark version of the CL? That way thin clients can build stark proofs and then executors will do the expensive wrapping to groth16?

**Pull request merged**